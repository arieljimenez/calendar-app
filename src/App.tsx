import React from 'react';
import { ThemeProvider } from 'theme-ui'

import { GlobalContextProvider } from './_contexts';

import { CalendarContainer } from './containers';
import theme from './configs/theme';

interface AppProps {}

function App({}: AppProps) {
  return (
    <GlobalContextProvider>
      <ThemeProvider theme={theme}>
        <CalendarContainer />
      </ThemeProvider>
    </GlobalContextProvider>
  );
}

export default App;
