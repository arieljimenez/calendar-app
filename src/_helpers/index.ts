interface getCellColorProps {
  idx: number;
  dayNumber: number;
}
/**
 * getCellColor
 * @param [number] idx - current cell index;
 * @param [number] dayNumber - current day
 * @returns [string] color
 */
export function getCellColor({ idx, dayNumber }: getCellColorProps): string {
  let color = 'black';

  // for the first and last week row, we need to dye other months days
  if ((idx < 6 && dayNumber > 7) || (idx > 27 && dayNumber < 7)) {
    color = 'gray';
  }
  // to know if is a weekend day and ignore if is from another month
  if ((idx % 7 == 0 || idx % 7 === 6) && color !== 'gray') {
    color = 'blue';
  }

  return color;
}


interface getMonthProps {
  date?: any;
  options?: {
    month?: string;
  }
}
/**
 * Returns the month of the giving date obj
 * @param {object} [date] - The date to work with.
 * @param {object} [options] - custom options
 * @param {string} [option.month] - the desired month format
 * @returns {string} month
 */
export function getMonth({ date = new Date(), options = {} }): string {
  const dateOptions = {
    month: 'long',
    ...options,
  }

  return new Intl.DateTimeFormat('en-US', dateOptions).format(date);
}

export function getMonthDays(month:string): number {
  const MAPPED_MONTHS: {[key:string]: number} = {
    january: 31,
    february: 28,
    march: 31,
    april: 30,
    may: 31,
    june: 30,
    july: 31,
    august: 31,
    september: 30,
    october: 31,
    november: 30,
    december: 31,
  }

  let days = MAPPED_MONTHS[month.toLowerCase()];

  // if month february and leap year return 29
  // solution https://twitter.com/colindecarlo/status/1360617963297112066?s=09
  return days;
}

/**
 *
 * @param keyName string value  to save the values under that name
 * @param payload object to save
 * @void
 */
export function setToLocalStore(keyName: string, payload: object): void {
  localStorage.setItem(keyName, JSON.stringify(payload));
}

/**
 * get the local stored values under the key
 * @param keyName - keyValue on in the store
 * @returns object or null
 */
export function getFromLocalStore(keyName: string): any | null {
  const value = localStorage.getItem(keyName);
  return value ? JSON.parse(value) : null;
}

/**
 *
 * @param keyName string value  to save the values under that name
 * @param payload object to save
 * @void
 */
export function delFromLocalStore(keyName: string): void {
  localStorage.removeItem(keyName);
}
