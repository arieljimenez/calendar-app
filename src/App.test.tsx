import * as React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import App from './App';

import { getMonth } from './_helpers';


describe('<Calendar />', () => {
  it('it renders current month name on the header', () => {
    const { getByText } = render(<App />);
    const currentMonth = getMonth({date: new Date()});
    const renderedMonth = getByText(currentMonth);
    expect(renderedMonth).toBeInTheDocument();
  });

  it('should match the snapshot', () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('<Modal />', () => {
    it('should render a modal when a day is clicked', () =>{
      const { getByText, queryByText } = render(<App />);

      let modalElement = queryByText(/New Appointment for/i);
      expect(modalElement).not.toBeInTheDocument();
      expect(modalElement).toBe(null);

      const cellElement = screen.getByTestId('calendar-cell-day-3');
      fireEvent.click(cellElement);

      modalElement = getByText(/New Appointment for/i);
      expect(modalElement).toBeInTheDocument();
    });

    it('should allow add a 30 character max description', () => {
      const { getByLabelText } = render(<App />);

      const cellElement = screen.getByTestId('calendar-cell-day-3');
      fireEvent.click(cellElement);

      const descInput = getByLabelText(/Description/i);
      expect(descInput.maxLength).toBe(30);
    });

    it('should allow to save a event without a city or desc', () => {
      const { getByLabelText, getByText, getAllByText } = render(<App />);

      const cellElement = screen.getByTestId('calendar-cell-day-3');
      fireEvent.click(cellElement);

      const cityInput = getByLabelText(/City/i);
      expect(cityInput.value).toBe('New York'); //by default

      const SaveBtn = getByText(/Save/i);

      fireEvent.click(SaveBtn);
      expect(getAllByText(/Required/i).length).toBe(1); // the desc is missing

      fireEvent.change(cityInput, { target: { value: ''}});
      expect(cityInput.value).toBe('');

      fireEvent.click(SaveBtn);
      expect(getAllByText(/Required/i).length).toBe(2);
    })
  });
});
