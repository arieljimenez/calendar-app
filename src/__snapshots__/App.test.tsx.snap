// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calendar /> should match the snapshot 1`] = `
<div
  className="css-ixhfdv"
>
  <div
    className="css-wzcygq"
  >
    <div
      className="currentMonth css-10hvtwb"
    >
      February
    </div>
    <div
      className="css-1kopen8"
    >
      <div
        className="css-1vz21y"
      >
        Sunday
      </div>
      <div
        className="css-1vz21y"
      >
        Monday
      </div>
      <div
        className="css-1vz21y"
      >
        Tuesday
      </div>
      <div
        className="css-1vz21y"
      >
        Wednesday
      </div>
      <div
        className="css-1vz21y"
      >
        Thursday
      </div>
      <div
        className="css-1vz21y"
      >
        Friday
      </div>
      <div
        className="css-1vz21y"
      >
        Saturday
      </div>
    </div>
  </div>
  <div
    className="css-vurnku"
  >
    <div
      className="css-x8mosh"
    >
      <div
        className="calendar-cell emtpy-cell css-uwdze5"
        data-testid="calendar-cell-day-0"
        onClick={[Function]}
      />
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-1"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-1"
          id="calendar-day-1"
        >
          1
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-2"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-2"
          id="calendar-day-2"
        >
          2
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-3"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-3"
          id="calendar-day-3"
        >
          3
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-4"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-4"
          id="calendar-day-4"
        >
          4
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-5"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-5"
          id="calendar-day-5"
        >
          5
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-6"
        onClick={[Function]}
      >
        <div
          className="css-17sdwej"
          data-testid="calendar-day-6"
          id="calendar-day-6"
        >
          6
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-7"
        onClick={[Function]}
      >
        <div
          className="css-17sdwej"
          data-testid="calendar-day-7"
          id="calendar-day-7"
        >
          7
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-8"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-8"
          id="calendar-day-8"
        >
          8
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-9"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-9"
          id="calendar-day-9"
        >
          9
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-10"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-10"
          id="calendar-day-10"
        >
          10
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-11"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-11"
          id="calendar-day-11"
        >
          11
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-12"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-12"
          id="calendar-day-12"
        >
          12
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-13"
        onClick={[Function]}
      >
        <div
          className="css-17sdwej"
          data-testid="calendar-day-13"
          id="calendar-day-13"
        >
          13
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-14"
        onClick={[Function]}
      >
        <div
          className="css-17sdwej"
          data-testid="calendar-day-14"
          id="calendar-day-14"
        >
          14
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-15"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-15"
          id="calendar-day-15"
        >
          15
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-16"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-16"
          id="calendar-day-16"
        >
          16
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-17"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-17"
          id="calendar-day-17"
        >
          17
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-18"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-18"
          id="calendar-day-18"
        >
          18
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-19"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-19"
          id="calendar-day-19"
        >
          19
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-20"
        onClick={[Function]}
      >
        <div
          className="css-17sdwej"
          data-testid="calendar-day-20"
          id="calendar-day-20"
        >
          20
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-21"
        onClick={[Function]}
      >
        <div
          className="css-17sdwej"
          data-testid="calendar-day-21"
          id="calendar-day-21"
        >
          21
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-22"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-22"
          id="calendar-day-22"
        >
          22
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-23"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-23"
          id="calendar-day-23"
        >
          23
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-24"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-24"
          id="calendar-day-24"
        >
          24
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-25"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-25"
          id="calendar-day-25"
        >
          25
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-26"
        onClick={[Function]}
      >
        <div
          className="css-1sy89y0"
          data-testid="calendar-day-26"
          id="calendar-day-26"
        >
          26
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-27"
        onClick={[Function]}
      >
        <div
          className="css-17sdwej"
          data-testid="calendar-day-27"
          id="calendar-day-27"
        >
          27
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell  css-uwdze5"
        data-testid="calendar-cell-day-28"
        onClick={[Function]}
      >
        <div
          className="css-17sdwej"
          data-testid="calendar-day-28"
          id="calendar-day-28"
        >
          28
        </div>
        <div
          className="css-anjba9"
        />
      </div>
      <div
        className="calendar-cell emtpy-cell css-uwdze5"
        data-testid="calendar-cell-day-0"
        onClick={[Function]}
      />
      <div
        className="calendar-cell emtpy-cell css-uwdze5"
        data-testid="calendar-cell-day-0"
        onClick={[Function]}
      />
      <div
        className="calendar-cell emtpy-cell css-uwdze5"
        data-testid="calendar-cell-day-0"
        onClick={[Function]}
      />
      <div
        className="calendar-cell emtpy-cell css-uwdze5"
        data-testid="calendar-cell-day-0"
        onClick={[Function]}
      />
      <div
        className="calendar-cell emtpy-cell css-uwdze5"
        data-testid="calendar-cell-day-0"
        onClick={[Function]}
      />
      <div
        className="calendar-cell emtpy-cell css-uwdze5"
        data-testid="calendar-cell-day-0"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;
